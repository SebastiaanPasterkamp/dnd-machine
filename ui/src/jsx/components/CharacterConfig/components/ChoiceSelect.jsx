import React from 'react';
import PropTypes from 'prop-types';
import {
    entries,
    filter,
    flow,
    map,
} from 'lodash/fp';

import LazyComponent from '../../LazyComponent.jsx';
import TabComponent from '../../TabComponent.jsx';

import CharacterConfig from '../CharacterConfig.jsx';
import MatchesFilters from '../utils/MatchesFilters';

export const ChoiceSelect = function({ options, filter: filters }) {
    const filtered = filter(
        (option) => !option.hidden && MatchesFilters(option, filters)
    )(options);

    return (
        <TabComponent tabConfig={filtered}>
            {flow(entries, map(
                ([index, option]) => (
                    <CharacterConfig
                        key={ index }
                        config={ [option] }
                    />
                )
            ))(filtered)}
        </TabComponent>
    );
};

ChoiceSelect.propTypes = {
    type: PropTypes.oneOf(['choice']).isRequired,
    options: PropTypes.arrayOf(
        PropTypes.shape({
            description: PropTypes.string,
        })
    ).isRequired,
    description: PropTypes.string,
};

export default ChoiceSelect;
