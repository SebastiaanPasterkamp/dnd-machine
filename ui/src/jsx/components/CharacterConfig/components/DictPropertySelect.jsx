import React from 'react';
import PropTypes from 'prop-types';
import { sprintf } from 'sprintf-js';
import MDReactComponent from 'markdown-react-js';
import {
    keys,
    reduce,
} from 'lodash/fp';

import CharacterEditorWrapper from '../hocs/CharacterEditorWrapper';

export const DictPropertySelect = function({ hidden, current })
{
    if (hidden || !current) {
        return null;
    }

    let value = current.description;
    const defaulted = reduce(
        (defaulted, key) => {
            if (!key.match(/_default$/)) {
                return defaulted;
            }
            const field = key.replace(/_default$/, '');
            return {
                [field]: current[key],
                ...defaulted,
            };
        },
        {}
    )(keys(current));
    try {
        value = sprintf(
            current.description || '',
            { ...defaulted, ...current }
        );
    } catch(err) {
        console.error(err, current);
    }

    return (
        <MDReactComponent text={value} />
    );
};

DictPropertySelect.propTypes = {
    type: PropTypes.oneOf(['dict']).isRequired,
    uuid: PropTypes.string.isRequired,
    current: PropTypes.object,
    dict: PropTypes.object,
    hidden: PropTypes.bool,
};

export default CharacterEditorWrapper(DictPropertySelect);
