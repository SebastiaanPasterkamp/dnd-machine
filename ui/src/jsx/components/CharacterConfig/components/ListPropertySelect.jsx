import React from 'react';
import PropTypes from 'prop-types';
import {
    concat,
    countBy,
    entries,
    every,
    find,
    filter,
    flow,
    forEach,
    get,
    includes,
    isArray,
    isNil,
    intersection,
    map,
    keys,
    some,
    without,
} from 'lodash/fp';

import {
    BaseTagContainer,
    Tag,
    TagBadge,
} from '../../BaseTagContainer';
import LazyComponent from '../../LazyComponent';
import SingleSelect from '../../SingleSelect';

import { memoize } from '../../../utils';

import CharacterEditorWrapper from '../hocs/CharacterEditorWrapper';
import ListsToItemsWrapper from '../../../hocs/ListsToItemsWrapper';

export class ListPropertySelect extends LazyComponent
{
    constructor(props) {
        super(props);
        this.state = {
            added: [],
            removed: [],
            showSelect: false,
            disabled: true,
        };
        this.memoize = memoize.bind(this);
    }

    static getDerivedStateFromProps(props, state) {
        const { add, limit, replace, current } = props;
        const { added, removed } = state;

        const showSelect = Boolean(
            (state.added.length - state.removed.length) < add
            || current.length < limit
        );
        const disabled = removed.length >= replace;

        if (showSelect !== state.showSelect
            || disabled !== state.disabled
        ) {
            return { showSelect, disabled };
        }
        return null;
    }

    componentDidMount() {
        const { given } = this.props;
        if (given.length) {
            this.onSetState();
        }
    }

    onSetState() {
        const { onChange, given } = this.props;
        const { added, removed } = this.state;
        onChange({
            added: concat(added, given),
            removed,
        });
    }

    onDelete = (value) => this.memoize(
        `delete-${value}`,
        () => {
            const { replace } = this.props;
            const { added, removed } = this.state;
            const state = { added, removed };
            if (includes(value, added)) {
                state.added = without([ value ], added);
            } else if (removed.length < replace) {
                state.removed = concat(removed, [value]);
            }
            this.setState(state, () => this.onSetState());
        }
    );

    onAdd = (value) => {
        const { add, limit, current } = this.props;
        const { added, removed } = this.state;

        if (limit) {
            if (current.length >= limit) {
                return false;
            }
        } else if (added.length >= (add + removed.length)) {
            return false;
        }

        const state = { added, removed };
        if (includes(value, removed)) {
            state.removed = without([ value ], removed);
        } else {
            state.added = concat(added, [value]);
        }
        this.setState(state, () => this.onSetState());
    }

    findItem(value, _default={label: value, color: 'bad'}) {
        const { items = [] } = this.props;
        const match = find({code: value}, items)
            || find({id: value}, items)
            || find({name: value}, items);

        if (!match && !_default) {
            return _default;
        }

        const item = match || {
            ..._default,
            code: value,
            label: value,
        };

        return {
            id: isNil(item.id) ? ( isNil(item.code) ? item.name : item.code ) : item.id,
            label: isNil(item.label) ? item.name : item.label,
            description: item.description,
        };
    }

    getValue() {
        const { given, current, replace } = this.props;
        const { added, removed } = this.state;

        const tags = concat(
            map(
                code => ({
                    ...this.findItem(code),
                    color: 'info',
                })
            )(added),
            map(
                code => ({
                    ...this.findItem(code),
                    color: 'good',
                    disabled: true,
                })
            )(given)
        );

        let upgrade = countBy('id', tags);
        let downgrade = countBy('id', removed);
        forEach(
            (count, code) => {
                upgrade[code] = (upgrade[code] || 0) - count;
            }
        )(downgrade);

        const disabled = (replace - removed.length) <= 0;
        return concat(
            tags,
            filter(null, map(
                code => {
                    if ((upgrade[code] || 0) > 0) {
                        upgrade[code] -= 1;
                        return null;
                    }
                    const _current = this.findItem(code, null);
                    if (!_current) {
                        return null;
                    }
                    return ({
                        ..._current,
                        color: disabled ? null : 'warning',
                        disabled,
                    });
                }
            )(without(removed, current)))
        );
    }

    matchesFilters(item, filters) {
        return flow(entries, every(
            ([path, cond]) => {
                if (path.match(/_(formula|default)$/)) {
                    return true;
                }
                if (path === 'or') {
                    return some(
                        option => this.matchesFilters(item, option)
                    )(cond);
                }
                if (path === 'and') {
                    return this.matchesFilters(item, cond);
                }
                if (path === 'not') {
                    return !this.matchesFilters(item, cond);
                }
                const value = get(path, item);
                return intersection(
                    isArray(value) ? value : [value],
                    isArray(cond) ? cond : [cond],
                ).length;
            }
        ))(filters);
    }

    renderSelect() {
        const {
            multiple, filter: filters, items, current, given,
            limit, add, replace,
        } = this.props;
        const { added, removed, showSelect } = this.state;

        if (!items.length) return null;

        if (!showSelect) {
            return null;
        }

        const values = [...given, ...current, ...added];
        const filtered = filter(
            item => (
                multiple
                || includes(
                    isNil(item.code) ? item.name : item.code,
                    removed
                )
                || !includes(
                    isNil(item.code) ? item.name : item.code,
                    values
                )
            ),
            filter(
                item => (
                    includes(
                        isNil(item.code) ? item.name : item.code,
                        removed
                    )
                    || this.matchesFilters(item, filters)
                ),
                items
            )
        );

        if (!filtered.length) return null;

        return (
            <SingleSelect
                emptyLabel="Add..."
                items={filtered}
                setState={this.onAdd}
            />
        );
    }

    render() {
        const { hidden } = this.props;
        if (hidden) {
            return null;
        }

        const tags = map(
            tag => ({
                ...tag,
                onDelete: this.onDelete(tag.id),
            })
        )(this.getValue());

        return (
            <BaseTagContainer>
                {this.renderSelect()}

                {flow(entries, map(([i, tag]) => (
                    <Tag
                        key={`tag-${i}`}
                        {...tag}
                    />
                )))(tags)}
            </BaseTagContainer>
        );
    }

};

ListPropertySelect.propTypes = {
    type: PropTypes.oneOf(['list']).isRequired,
    onChange: PropTypes.func.isRequired,
    items: PropTypes.arrayOf(PropTypes.object),
    given: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number
        ])
    ),
    current: PropTypes.array,
    limit: PropTypes.number,
    add: PropTypes.number,
    replace: PropTypes.number,
    filter: PropTypes.object,
    multiple: PropTypes.bool,
    hidden: PropTypes.bool,
};

ListPropertySelect.defaultProps = {
    given: [],
    current: [],
    limit: 0,
    add: 0,
    replace: 0,
    multiple: false,
    filter: {},
    items: [],
};

export default ListsToItemsWrapper(
    CharacterEditorWrapper(
        ListPropertySelect,
        {
            current: true,
        }
    ),
    'items'
);
