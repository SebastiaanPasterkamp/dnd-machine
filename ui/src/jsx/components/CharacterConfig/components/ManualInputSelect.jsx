import React from 'react';
import PropTypes from 'prop-types';

import InputField from '../../InputField';
import MarkdownTextField from '../../MarkdownTextField';

import CharacterEditorWrapper from '../hocs/CharacterEditorWrapper';

export class ManualInputSelect extends React.Component
{
    constructor(props) {
        super(props);
        this.onSetState = this.onSetState.bind(this);
    }

    onSetState(current) {
        const { setState } = this.props;
        setState({ current });
    };

    render() {
        const { current, markup, placeholder } = this.props;

        if(markup) {
            return (
                <MarkdownTextField
                    placeholder={placeholder}
                    value={current}
                    rows={5}
                    setState={this.onSetState}
                />
            );
        }

        return (
            <InputField
                placeholder={placeholder}
                value={current}
                setState={this.onSetState}
            />
        );
    }
};

ManualInputSelect.propTypes = {
    type: PropTypes.oneOf(['manual']).isRequired,
    uuid: PropTypes.string.isRequired,
    setState: PropTypes.func.isRequired,
    current: PropTypes.string,
    placeholder: PropTypes.string,
    markup: PropTypes.bool,
};

ManualInputSelect.defaultProps = {
    current: '',
    placeholder: '',
    markup: false,
};

export default CharacterEditorWrapper(ManualInputSelect);
