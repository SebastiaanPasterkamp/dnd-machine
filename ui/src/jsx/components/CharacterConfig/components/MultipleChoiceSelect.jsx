import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import TagContainer from '../../TagContainer';

import CharacterConfig from '../CharacterConfig';
import CharacterEditorWrapper from '../hocs/CharacterEditorWrapper';

export class MultipleChoiceSelect extends React.Component
{
    constructor(props) {
        super(props);
        this.state = {
            added: [],
            removed: [],
            filtered: [],
            showSelect: false,
            disabled: true,
        };
    }

    static getDerivedStateFromProps(props, state) {
        const { add, limit, replace } = props;
        const { added, removed } = state;
        const current = MultipleChoiceSelect.getCurrent(props, state);

        const showSelect = Boolean(
            (added.length - removed.length) < add
            || current.length < limit
        );
        const disabled = removed.length >= replace;

        if (showSelect !== state.showSelect
            || disabled != state.disabled
        ) {
            return { showSelect, disabled };
        }
        return null;
    }

    onAdd = (label) => {
        const { limit, add, replace } = this.props;
        const { added, removed, filtered } = this.state;

        const current = MultipleChoiceSelect.getCurrent(
            this.props,
            this.state
        );

        if (limit) {
            if (current.length >= limit) {
                return false;
            }
        } else if (added.length >= (add + removed.length)) {
            return false;
        }

        const state = { added, removed, filtered };
        if (_.includes(removed, label)) {
            state.removed = _.without(removed, label);
        } else if (added.length < (add + removed.length)) {
            state.filtered = _.without(filtered, label);
            state.added = _.concat(added, [label]);
        }
        this.setState(state);
    }

    onDelete = (label) => {
        const { limit, add, replace } = this.props;
        const { added, removed, filtered } = this.state;
        const state = { added, removed, filtered };

        if (_.includes(added, label)) {
            state.added = _.without(added, label);
            state.filtered = _.concat(removed, [label]);
        } else if (removed.length < replace) {
            state.removed = _.concat(removed, [label]);
        }
        this.setState(state);
    }

    onSetState = () => null;

    static getCurrent(props, state) {
        const { options, getCurrent } = props;
        const { added, removed, filtered } = state;

        const current = _.chain(options)
            .filter(option => {
                if (_.includes(added, option.label)) {
                    return true;
                }
                if (_.includes(filtered, option.label)) {
                    return false;
                }
                if (_.includes(removed, option.label)) {
                    return false;
                }
                const path = (
                    _.get(option, 'path')
                    || _.get(option, ['config', 0, 'path'])
                );
                return getCurrent(path);
            })
            .map(option => option.label)
            .value();
        return current;
    }

    render() {
        const { options, limit, add, replace } = this.props;
        const {
            added, removed, filtered, showSelect, disabled,
        } = this.state;

        const current = MultipleChoiceSelect.getCurrent(
            this.props,
            this.state
        );
        const itemsDisabled = disabled || removed.length >= replace;
        const items = _.chain(options)
            .filter(option => !option.hidden)
            .map(option => {
                const isNew = _.includes(added, option.label);
                return {
                    code: option.label,
                    label: option.label,
                    color: isNew ? 'info' : 'warning',
                    disabled: !isNew && itemsDisabled,
                };
            })
            .value();

        return (
            <div>
                <TagContainer
                    value={current}
                    items={items}
                    onAdd={ this.onAdd }
                    onDelete={ this.onDelete }
                    setState={ this.onSetState }
                    showSelect={ showSelect }
                />
                {_.map(
                    options,
                    (config, i) => _.includes(current, config.label) ? (
                        <CharacterConfig
                            key={i}
                            config={ [config] }
                        />
                    ) : null
                )}
            </div>
        );
    }
};

MultipleChoiceSelect.defaultProps = {
    limit: 0,
    add: 0,
    replace: 0,
};

MultipleChoiceSelect.propTypes = {
    type: PropTypes.oneOf(['multichoice']).isRequired,
    options: PropTypes.arrayOf(PropTypes.object).isRequired,
    getCurrent: PropTypes.func.isRequired,
    description: PropTypes.string,
    limit: PropTypes.number,
    add: PropTypes.number,
    replace: PropTypes.number,
};

export default CharacterEditorWrapper(
    MultipleChoiceSelect,
    {
        character: true,
    }
);
