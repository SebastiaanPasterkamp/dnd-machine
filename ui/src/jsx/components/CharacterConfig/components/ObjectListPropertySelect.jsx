import React from 'react';
import PropTypes from 'prop-types';
import {
    concat,
    countBy,
    find,
    findIndex,
    filter,
    forEach,
    includes,
    isPlainObject,
    map,
    omitBy,
    sumBy,
    uniq,
    without,
} from 'lodash/fp';

import {
    BaseTagContainer,
    Tag,
    TagBadge,
} from '../../BaseTagContainer';
import SingleSelect from '../../SingleSelect';

import { memoize } from '../../../utils';

import CharacterEditorWrapper from '../hocs/CharacterEditorWrapper';
import ListsToItemsWrapper from '../../../hocs/ListsToItemsWrapper';
import MatchesFilters from '../utils/MatchesFilters';

export class ObjectListPropertySelect extends React.Component
{
    constructor(props) {
        super(props);
        this.state = {
            showSelect: false,
            disabled: true,
        };
        this.memoize = memoize.bind(this);
        this.onAdd = this.onAdd.bind(this);
        this.onDelete = this.onDelete.bind(this);
    }

    static getDerivedStateFromProps(props, state) {
        const { added, removed, given, current, add, limit, replace } = props;
        const cntAdded = sumBy(a => a.count || 1, added);
        const cntCurrent = sumBy(c => c.count || 1, current);
        const cntRemoved = sumBy(r => r.count || 1, removed);

        const showSelect = (cntAdded - cntRemoved) < add || cntCurrent < limit;
        const disabled = cntRemoved >= replace;

        if (showSelect !== state.showSelect || disabled !== state.disabled) {
            return { showSelect, disabled };
        }
        return null;
    }

    getNeedle(item) {
        const { type, id } = item || {};
        return omitBy(
            (value) => value === undefined
        )({type, id});
    }

    increase(item, set) {
        const { multiple } = this.props;
        const needle = this.getNeedle(item);

        const idx = findIndex(needle, set);
        if (idx >= 0) {
            return [
                ...set.slice(0, idx),
                {...set[idx], ...item, count: multiple ? set[idx].count + 1 : undefined},
                ...set.slice(idx + 1),
            ];
        }

        return [...set, {...item, count: multiple ? 1 : undefined}];
    }

    decrease(item, set) {
        const { multiple } = this.props;
        const needle = this.getNeedle(item);

        const idx = findIndex(needle, set);
        if (multiple && set[idx].count > 1) {
            return [
                ...set.slice(0, idx),
                {...set[idx], ...item, count: set[idx].count - 1},
                ...set.slice(idx + 1),
            ];
        }

        return [
            ...set.slice(0, idx),
            ...set.slice(idx + 1),
        ];
    }

    onDelete(item) {
        const { type, id } = item;
        const key = `delete-${type}-${id}`;
        const needle = this.getNeedle(item);

        return this.memoize(key, () => {
            const { added, removed, replace, setState } = this.props;
            const cntRemoved = sumBy(r => r.count || 1, removed);

            const state = { added, removed };
            if (find(needle, added)) {
                state.added = this.decrease(item, added);
            } else if (cntRemoved < replace) {
                state.removed = this.increase(item, removed);
            }

            setState(state);
        });
    }

    onAdd(item) {
        const { items, added, removed, setState } = this.props;
        const needle = this.getNeedle(item);

        const state = { added, removed };
        if (find(needle, removed)) {
            state.removed = this.decrease(item, removed);
        } else {
            state.added = this.increase(item, added);
        }

        setState(state);
    }

    getValue() {
        const {
            added, removed, items, current, given, replace, uuid
        } = this.props;
        const { disabled } = this.state;

        return map(
            (item) => {
                const needle = this.getNeedle(item);

                const isAdded = find(needle, added) !== undefined;
                const isGiven = find(needle, given) !== undefined;
                const isRemoved = find(needle, removed) !== undefined;

                return {
                    ...item,
                    color: isAdded ? 'info' : (
                        isRemoved ? 'bad' : (
                            isGiven ? 'good' : (
                                disabled && !isAdded ? null : 'warning'
                            )
                        )
                    ),
                    disabled: (disabled || isGiven) && !isAdded,
                };
            }
        )(current);
    }

    renderSelect() {
        const {
            multiple, items, current, removed,
            filter: filters,
        } = this.props;
        const { showSelect } = this.state;

        if (!items.length) return null;

        if (!showSelect) {
            return null;
        }

        const filtered = filter(
            item => {
                const needle = this.getNeedle(item);
                if (find(needle, removed)) {
                    return true;
                }
                if (!MatchesFilters(item, filters)) {
                    return false;
                }
                if (multiple) {
                    return true;
                }
                if (find(needle, current)) {
                    return false;
                }
                return true;
            }
        )(items);

        if (!filtered.length) return null;

        return (
            <SingleSelect
                emptyLabel="Add..."
                items={filtered}
                setState={this.onAdd}
                objects={true}
                filterable={true}
            />
        );
    }

    render() {
        const { hidden } = this.props;

        if (hidden) {
            return null;
        }

        return (
            <BaseTagContainer>
                {this.renderSelect()}

                {map(
                    tag => (
                        <Tag
                            key={`tag-${tag.type}-${tag.id}`}
                            onDelete={this.onDelete(tag)}
                            {...tag}
                        >
                            {tag.count > 1 ? (
                                <TagBadge>
                                    &times; {tag.count}
                                </TagBadge>
                            ) : null}
                        </Tag>
                    )
                )(this.getValue())}
            </BaseTagContainer>
        );
    }
};

const itemType = {
    id: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
    ]).isRequired,
    type: PropTypes.string,
    name: PropTypes.string,
    description: PropTypes.string,
    count: PropTypes.number,
};

ObjectListPropertySelect.propTypes = {
    type: PropTypes.oneOf(['objectlist']).isRequired,
    uuid: PropTypes.string.isRequired,
    setState: PropTypes.func.isRequired,
    items: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
        ]),
        name: PropTypes.string,
        description: PropTypes.string,
    })),
    given: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.shape(itemType),
            PropTypes.string,
            PropTypes.number,
        ]),
    ),
    current: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.shape(itemType),
            PropTypes.string,
            PropTypes.number,
        ]),
    ),
    limit: PropTypes.number,
    added: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.shape(itemType),
            PropTypes.string,
            PropTypes.number,
        ]),
    ),
    removed: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.shape(itemType),
            PropTypes.string,
            PropTypes.number,
        ]),
    ),
    add: PropTypes.number,
    replace: PropTypes.number,
    filter: PropTypes.arrayOf(
        PropTypes.object
    ),
    multiple: PropTypes.bool,
    hidden: PropTypes.bool,
};

ObjectListPropertySelect.defaultProps = {
    added: [],
    removed: [],
    given: [],
    current: [],
    limit: 0,
    add: 0,
    replace: 0,
    multiple: false,
    filter: [],
    items: [],
    hidden: false,
};

export default ListsToItemsWrapper(
    CharacterEditorWrapper(ObjectListPropertySelect),
    'items'
);
