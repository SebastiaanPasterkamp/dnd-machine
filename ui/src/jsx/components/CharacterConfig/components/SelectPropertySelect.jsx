import React from 'react';
import PropTypes from 'prop-types';

import LazyComponent from '../../LazyComponent';
import SingleSelect from '../../SingleSelect';

import CharacterEditorWrapper from '../hocs/CharacterEditorWrapper';
import ListsToItemsWrapper from '../../../hocs/ListsToItemsWrapper';

export class SelectPropertySelect extends LazyComponent
{
    constructor(props) {
        super(props);
        this.onChange = this.onChange.bind(this);
    }

    componentDidMount() {
        const { current, onChange } = this.props;
        if (current !== undefined) {
            onChange(this.props.current);
        }
    }

    onChange(current) {
        const { onChange, setState } = this.props;
        setState({ current });
        onChange(current);
    }

    render() {
        const { hidden, items, current } = this.props;

        if (hidden) {
            return null;
        }

        return (
            <SingleSelect
                className="small"
                items={items}
                setState={this.onChange}
                selected={current}
                emptyLabel="Please select"
            />
        );
    }
};

SelectPropertySelect.propTypes = {
    type: PropTypes.oneOf(['select']).isRequired,
    uuid: PropTypes.string.isRequired,
    setState: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired,
    items: PropTypes.arrayOf(PropTypes.object).isRequired,
    current: PropTypes.any,
    hidden: PropTypes.bool,
};

export default ListsToItemsWrapper(
    CharacterEditorWrapper(
        SelectPropertySelect,
        {
            current: true,
        }
    ),
    'items'
);
