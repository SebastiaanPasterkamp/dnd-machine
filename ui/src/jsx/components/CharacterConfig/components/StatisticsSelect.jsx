import React from 'react';
import Reflux from 'reflux';
import {
    forEach,
    isEqual,
    uniqueId,
} from 'lodash/fp';

import actions from '../actions/CharacterEditorActions.jsx';
import ListDataWrapper from '../../../hocs/ListDataWrapper.jsx';
import store from '../stores/CharacterEditorStore.jsx';

import StatsBlock from '../../StatsBlock.jsx';

class StatisticsSelect extends Reflux.Component
{
    constructor(props) {
        super(props);
        this.store = store;
        this.storeKeys = ['character'];
        this._id = uniqueId();
    }

    componentWillUnmount() {
        super.componentWillUnmount.call(this);
        const { statistics } = this.props;

        actions.removeChange(`${ this._id }-bare`);
        actions.removeChange(`${ this._id }-base`);
        actions.removeChange(`${ this._id }-modifiers`);

        forEach(({ code }) => actions.removeChange(
            `${ this._id }-bonus-${ code }`
        ))(statistics);
    }

    onStatisticsChange = ({ bare, base, modifiers }) => {
        const {
            character: {
                statistics = {},
            },
        } = this.state;

        if (!isEqual(bare, statistics.bare)) {
            actions.addChange(
                'statistics.bare',
                bare,
                `${ this._id }-bare`,
                { type: 'dict' },
            );
        }
        if (!isEqual(base, statistics.base)) {
            actions.addChange(
                'statistics.base',
                base,
                `${ this._id }-base`,
                { type: 'dict' },
            );
        }
        if (!isEqual(modifiers, statistics.modifiers)) {
            actions.addChange(
                'statistics.modifiers',
                modifiers,
                `${ this._id }-modifiers`,
                { type: 'dict' },
            );
        }
    }

    onBonusChange = (bonus) => {
        const { statistics } = this.props;

        forEach(({ code }) => {
            if(code in bonus) {
                actions.addChange(
                    `statistics.bonus.${code}`,
                    {
                        removed: [],
                        added: bonus[code],
                    },
                    `${ this._id }-bonus-${ code }`,
                    {
                        type: 'list',
                        multiple: true,
                    },
                );
            } else {
                actions.removeChange(
                    `${ this._id }-bonus-${ code }`
                );
            }
        })(statistics);
    }

    render() {
        const {
            onChange, getCurrent, current, limit,
            ...props
        } = this.props;
        const {
            character: {
                statistics = {},
            }
        } = this.state;

        return (
            <StatsBlock
                editBase={false}
                {...props}
                {...statistics}
                increase={ limit }
                setState={ this.onStatisticsChange }
                bonusChange={ this.onBonusChange }
            />
        );
    }
};

StatisticsSelect.defaultProps = {
    statistics: [],
};

export default ListDataWrapper(
    StatisticsSelect,
    ['statistics'],
    'items'
);
