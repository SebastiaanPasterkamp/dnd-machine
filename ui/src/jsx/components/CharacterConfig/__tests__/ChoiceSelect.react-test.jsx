import React from 'react';
import { mount } from 'enzyme';

import { mockedApi, statistics } from '../../../../../tests/__mocks__';

import { ChoiceSelect } from '../components/ChoiceSelect';

describe('Component: ChoiceSelect', () => {
    const fullProps = {
        type: 'choice',
        uuid: 'mocked-uuid-1',
        selected: 'mocked-uuid-2',
        options: [{
            type: 'ability_score',
            uuid: 'mocked-uuid-2',
            label: 'Choice A',
            limit: 1,
        }, {
            type: 'ability_score',
            uuid: 'mocked-uuid-3',
            label: 'Choice B',
            description: 'Some *text* here',
            limit: 2,
        }],
    };

    beforeAll(() => {
        fetch.mockImplementation( mockedApi({
            statistics,
        }) );
    });

    afterAll(() => fetch.resetMocks());

    it('should render with minimum props', () => {
        const setState = jest.fn();
        const wrapper = mount(
            <ChoiceSelect
                setState={setState}
                {...fullProps}
                options={[]}
            />
        );

        expect(wrapper).toMatchSnapshot();
        expect(setState).not.toBeCalled();
    });

    it('should render with all props', () => {
        const setState = jest.fn();
        const wrapper = mount(
            <ChoiceSelect
                setState={setState}
                {...fullProps}
                selected="mocked-uuid-3"
            />
        );

        expect(wrapper).toMatchSnapshot();
        expect(setState).not.toBeCalled();
    });

    it('should emit the first tab if nothing is selected yet', () => {
        const setState = jest.fn();
        const wrapper = mount(
            <ChoiceSelect
                setState={setState}
                {...fullProps}
                selected={undefined}
            />
        );

        expect(setState).toBeCalledWith({
            selected: fullProps.options[0].uuid,
        });
    });

    it('should emit changes while switching tabs', () => {
        const setState = jest.fn();
        const wrapper = mount(
            <ChoiceSelect
                setState={setState}
                {...fullProps}
            />
        );

        wrapper
            .find('.tab-component__tab a')
            .at(1)
            .simulate('click');

        expect(setState).toBeCalledWith({
            selected: fullProps.options[1].uuid,
        });

        wrapper.setProps({
            selected: fullProps.options[1].uuid,
        });

        wrapper
            .find('.tab-component__tab a')
            .at(0)
            .simulate('click');

        expect(setState).toBeCalledWith({
            selected: fullProps.options[0].uuid,
        });
        wrapper.setProps({
            selected: fullProps.options[0].uuid,
        });
    });
});
