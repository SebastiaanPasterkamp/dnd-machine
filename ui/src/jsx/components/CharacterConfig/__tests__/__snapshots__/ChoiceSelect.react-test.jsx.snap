// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ChoiceSelect should render with all props 1`] = `
<ChoiceSelect
  options={
    Array [
      Object {
        "name": "Choice A",
        "path": "some.path",
        "type": "value",
        "uuid": "mocked-uuid-2",
        "value": "Option one",
      },
      Object {
        "description": "Some *text* here",
        "name": "Choice B",
        "path": "some.path",
        "type": "value",
        "uuid": "mocked-uuid-3",
        "value": "Option two",
      },
    ]
  }
  selected="mocked-uuid-3"
  setState={[Function]}
  type="choice"
  uuid="mocked-uuid-1"
>
  <TabComponent
    activeTab={1}
    onTabChange={[Function]}
    tabConfig={
      Array [
        Object {
          "name": "Choice A",
          "path": "some.path",
          "type": "value",
          "uuid": "mocked-uuid-2",
          "value": "Option one",
        },
        Object {
          "description": "Some *text* here",
          "name": "Choice B",
          "path": "some.path",
          "type": "value",
          "uuid": "mocked-uuid-3",
          "value": "Option two",
        },
      ]
    }
  >
    <div
      className="nice-tabs-wrapper tab-component"
    >
      <ul
        className="nice-tabs bold"
      >
        <Tab
          active={false}
          color={null}
          disabled={false}
          icon={null}
          key="tab-0/.$mocked-uuid-2"
          name="Choice A"
          onSelect={[Function]}
          path="some.path"
          type="value"
          uuid="mocked-uuid-2"
          value="Option one"
        >
          <li
            className="tab-component__tab"
          >
            <a
              className="cursor-pointer"
              onClick={[Function]}
            >
              Choice A
            </a>
          </li>
        </Tab>
        <Tab
          active={true}
          color={null}
          description="Some *text* here"
          disabled={false}
          icon={null}
          key="tab-1/.$mocked-uuid-3"
          name="Choice B"
          onSelect={[Function]}
          path="some.path"
          type="value"
          uuid="mocked-uuid-3"
          value="Option two"
        >
          <li
            className="tab-component__tab current"
          >
            <a
              className="cursor-pointer"
              onClick={null}
            >
              Choice B
            </a>
          </li>
        </Tab>
      </ul>
      <ul
        className="nice-tab-content"
      >
        <li
          className="current"
          key=".$mocked-uuid-3"
        >
          <CharacterConfig
            config={
              Array [
                Object {
                  "description": "Some *text* here",
                  "name": null,
                  "path": "some.path",
                  "type": "value",
                  "uuid": "mocked-uuid-3",
                  "value": "Option two",
                },
              ]
            }
            key="mocked-uuid-3"
            type="config"
            uuid="mocked-uuid-3"
          >
            <FormGroup
              key="mocked-uuid-3"
              label={null}
            >
              <div
                className="nice-form-group"
              >
                <label />
                <MDReactComponent
                  text="Some *text* here"
                >
                  <span
                    key="mdrct-html-0"
                  >
                    <p
                      key="mdrct-p-0"
                    >
                      Some 
                      <em
                        key="mdrct-em-1"
                      >
                        text
                      </em>
                       here
                    </p>
                  </span>
                </MDReactComponent>
                <CharacterEditorComponent
                  path="some.path"
                  type="value"
                  uuid="mocked-uuid-3"
                  value="Option two"
                >
                  <ValuePropertySelect
                    hidden={false}
                    onSave={[Function]}
                    onUpdate={[Function]}
                    setState={[Function]}
                    type="value"
                    uuid="mocked-uuid-3"
                    value="Option two"
                  >
                    <MDReactComponent
                      text="Option two"
                    >
                      <span
                        key="mdrct-html-0"
                      >
                        <p
                          key="mdrct-p-0"
                        >
                          Option two
                        </p>
                      </span>
                    </MDReactComponent>
                  </ValuePropertySelect>
                </CharacterEditorComponent>
              </div>
            </FormGroup>
          </CharacterConfig>
        </li>
      </ul>
    </div>
  </TabComponent>
</ChoiceSelect>
`;

exports[`Component: ChoiceSelect should render with minimum props 1`] = `
<ChoiceSelect
  options={Array []}
  selected="mocked-uuid-2"
  setState={[Function]}
  type="choice"
  uuid="mocked-uuid-1"
/>
`;
